Source/WebCore/Modules/url-pattern/URLPatternParser.cpp

```diff
@@ -322,13 +325,220 @@ static String escapeRegexStringForCharacters(std::span<const CharacterType> char
+
+        if (!needsGrouping && hasCustomName
+            && part.type == PartType::SegmentWildcard && part.modifier == Modifier::None
+            && nextPart && nextPart->prefix.isEmpty() && nextPart->suffix.isEmpty()) {
+            if (nextPart->type == PartType::FixedText)
+                needsGrouping = isValidNameCodepoint(*StringView(nextPart->value).codePoints().begin(), IsFirst::No);
+            else
+                needsGrouping = !nextPart->name.isEmpty() && isASCIIDigit(nextPart->name[0]);
+        }
+
+        if (!needsGrouping && part.prefix.isEmpty() && previousPart && previousPart->type == PartType::FixedText) {
+            if (options.prefixCodepoint.length() == 1
+                && options.prefixCodepoint.startsWith(*StringView(nextPart->value).codePoints().codePointAt(nextPart->value.length() - 1)))
+                needsGrouping = true;
+        }
+
+        ASSERT(!part.name.isEmpty());
+
+        if (needsGrouping)
+            result.append('{');
+
+        result.append(escapePatternString(part.prefix));
+
+        if (hasCustomName)
+            result.append(':', part.name);
+
+        if (part.type == PartType::Regexp)
+            result.append('(', part.value, ')');
+        else if (part.type == PartType::SegmentWildcard && !hasCustomName)
+            result.append('(', generateSegmentWildcardRegexp(options), ')');
+        else if (part.type == PartType::FullWildcard) {
+            if (!hasCustomName
+                && (!previousPart || previousPart->type == PartType::FixedText || previousPart->modifier != Modifier::None
+                    || needsGrouping || !part.prefix.isEmpty()))
+                result.append('*');
+            else
+                result.append("(.*)"_s);
+        }
+
+        if (part.type == PartType::SegmentWildcard && hasCustomName && !part.suffix.isEmpty() && isValidNameCodepoint(*StringView(part.suffix).codePoints().begin(), IsFirst::Yes))
+            result.append('\\');
+
+        result.append(escapePatternString(part.suffix));
+
+        if (needsGrouping)
+            result.append('}');
+
+        result.append(convertModifierToString(part.modifier));
+    }
+
+    return result.toString();
+}
+
+template<typename CharacterType>
+static String escapePatternStringForCharacters(std::span<const CharacterType> characters)
+{
+    static constexpr std::array escapeCharacters { '+', '*', '?', ':', '(', ')', '\\', '{', '}' }; // NOLINT
+
+    StringBuilder result;
+    result.reserveCapacity(characters.size());
+
+    for (auto character : characters) {
+        if (std::find(escapeCharacters.begin(), escapeCharacters.end(), character) != escapeCharacters.end())
+            result.append('\\');
+
+        result.append(character);
+    }
+
+    return result.toString();
+}
+
+// https://urlpattern.spec.whatwg.org/#escape-a-pattern-string
+String escapePatternString(StringView input)
+{
+    // FIXME: Add check to see that input only contains ASCII
```

Comment:
Do you mean "ensure that input only contains ASCII based on spec"?
I think we need to add that "after the parser (or tokenizer) knows to filter non-ASCII input".